x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/aggregation/CompositeAggregationDataExtractor.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+package org.elasticsearch.xpack.ml.datafeed.extractor.aggregation;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.elasticsearch.action.ActionRequestBuilder;
+import org.elasticsearch.action.search.SearchRequest;
+import org.elasticsearch.action.search.SearchResponse;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.search.aggregations.Aggregations;
+import org.elasticsearch.search.aggregations.bucket.composite.CompositeAggregation;
+import org.elasticsearch.search.aggregations.bucket.composite.CompositeAggregationBuilder;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.xpack.core.ClientHelper;
+import org.elasticsearch.xpack.core.ml.datafeed.extractor.DataExtractor;
+import org.elasticsearch.xpack.core.ml.datafeed.extractor.ExtractorUtils;
+import org.elasticsearch.xpack.ml.datafeed.DatafeedTimingStatsReporter;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * An implementation that extracts data from elasticsearch using search with composite aggregations on a client.
+ * The first time {@link #next()} is called, the search is executed. All the aggregated buckets from the composite agg are then
+ * returned. Subsequent calls to {@link #next()} execute additional searches, moving forward the composite agg `afterKey`.
+ *
+ * It's like scroll, but with aggs.
+ *
+ * It is not thread-safe, we reuse underlying objects without synchronization (like a pre-constructed composite agg object)
+ */
+class CompositeAggregationDataExtractor implements DataExtractor {
+
+    private static final Logger LOGGER = LogManager.getLogger(CompositeAggregationDataExtractor.class);
+
+    private final AtomicReference<Map<String, Object>> afterKey = new AtomicReference<>(null);
```

Comment:
I think we could have this simply `volatile`.

Reply:
Probably, it also doesn't really need to be volatile (a la is_cancelled)